### PR TITLE
feat: selfdestruct check

### PR DESCRIPTION
**Pull request by gzeoneth**
This PR add a SELFDESTRUCT detection check to contracts touched by the proposal using logic from https://github.com/MrLuit/selfdestruct-detect

edit: also detect DELEGATECALL, and try to differentiate eoa vs empty addr that might have code later using account nonce; note that it would flag all proxy as warning tho due to DELEGATECALL

same as https://github.com/ArbitrumFoundation/governance-seatbelt/pull/9

### CODE REVIEW

**Comment by mds1** [comment]
Thanks for this! I should be able to review it tomorrow. One thought is it might be nice to also check for delegatecalls, since they can be used to sneak by a selfdestruct (and maybe this already checks for that)

**Comment by gzeoneth** [comment]
> Thanks for this! I should be able to review it tomorrow. One thought is it might be nice to also check for delegatecalls, since they can be used to sneak by a selfdestruct (and maybe this already checks for that)

yes I forgot to update the PR description, this also detect DELEGATECALL, and try to differentiate eoa vs empty addr that might have code later using account nonce; it would flag all proxy as warning tho due to DELEGATECALL but I think its fine

File: checks/check-targets-no-selfdestruct.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * For a given address, check if it's an EOA, a safe contract, or a contract contain selfdestruct
+ */
+async function checkNoSelfdestruct(
+  sim: TenderlySimulation,
+  addr: string,
+  provider: JsonRpcProvider
+): Promise<'safe' | 'eoa' | 'empty' | 'selfdestruct' | 'delegatecall'> {
+
+  const code = await provider.getCode(addr)
+  const nonce = await provider.getTransactionCount(addr)
```
Comment (mds1) [inline checks/check-targets-no-selfdestruct.ts:74]: We can dispatch these promises at the same time
```suggestion
  const [code, nonce] = await Promise.all([provider.getCode(addr), provider.getTransactionCount(addr)])
```

**Review by mds1 — COMMENTED** [review]
Looks great overall, two small comments which should apply to the arbitrum version too

File: checks/check-targets-no-selfdestruct.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import { bullet, toAddressLink } from '../presentation/report'
+import { ProposalCheck, TenderlySimulation } from '../types'
+
+/**
+ * Check all targets with code if they contain selfdestruct.
+ */
+export const checkTargetsNoSelfdestruct: ProposalCheck = {
+  name: 'Check all targets do not contain selfdestruct',
+  async checkProposal(proposal, sim, deps) {
+    const uniqueTargets = proposal.targets.filter((addr, i, targets) => targets.indexOf(addr) === i)
```
[the remark this answers is not in the document]
Comment (mds1) [reply checks/check-targets-no-selfdestruct.ts:11]: (Also just a heads up I pushed a prettier commit to your branch, was going to merge it but then noticed this)

[the remark this answers is not in the document]
Comment (gzeoneth) [reply checks/check-targets-no-selfdestruct.ts:11]: it also make sense to be able to supply a list of whitelisted addresses but haven't figured out where is the best place to do so

[the remark this answers is not in the document]
Comment (mds1) [reply checks/check-targets-no-selfdestruct.ts:11]: That's an interesting idea—are you thinking e.g. arbitrum system contracts that show up in proposals frequently and therefore always have delegatecalls, but don't add much value to the proposal check?

Maybe a bit clunky, but we could use an env var called something like `CHECK_EXCLUSIONS` or similar, so the CI file would look like this:

```yml
jobs:
  check-proposals:
    strategy:
      # Set fail-fast to false to ensure all Matrix runs complete even if one fails.
      fail-fast: false
      matrix:
        include:
          - DAO_NAME: 'Uniswap'
            GOVERNOR_ADDRESS: '0x408ED6354d4973f66138C91495F2f2FCbd8724C3'
          - DAO_NAME: 'Compound'
            GOVERNOR_ADDRESS: '0xc0Da02939E1441F497fd74F78cE7Decb17B66529'
            CHECK_EXCLUSIONS: |
              checkTargetsNoSelfdestruct:[0x123, 0xdef]
              otherCheckName:{thisOne: "takesAnObject"}
```

Then we parse out the check names and pass everything provided to the `checkProposal` `deps` arg

**Review by mds1 — APPROVED** [review]

**Pull request merged**